chains/substrate/client/client.go

```diff
@@ -0,0 +1,141 @@
+// The Licensed Work is (c) 2022 Sygma
+// SPDX-License-Identifier: BUSL-1.1
+
+package client
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types/codec"
+
+	"github.com/ChainSafe/sygma-relayer/chains/substrate/connection"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/signature"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+	"github.com/rs/zerolog/log"
+)
+
+type SubstrateClient struct {
+	key       *signature.KeyringPair // Keyring used for signing
+	nonceLock sync.Mutex             // Locks nonce for updates
+	nonce     types.U32              // Latest account nonce
+
+	Conn    *connection.Connection
+	ChainID *big.Int
+}
+
+func NewSubstrateClient(conn *connection.Connection, key *signature.KeyringPair, chainID *big.Int) *SubstrateClient {
+	return &SubstrateClient{
+		key:     key,
+		Conn:    conn,
+		ChainID: chainID,
+	}
+}
+
+// Transact constructs and submits an extrinsic to call the method with the given arguments.
+// All args are passed directly into GSRPC. GSRPC types are recommended to avoid serialization inconsistencies.
+func (c *SubstrateClient) Transact(method string, args ...interface{}) (*types.Hash, error) {
+	log.Debug().Msgf("Submitting substrate call... method %s, sender %s", method, c.key.Address)
+
+	// Create call and extrinsic
+	meta := c.Conn.GetMetadata()
+	call, err := types.NewCall(
+		&meta,
+		method,
+		args...,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct call: %w", err)
+	}
+
+	ext := types.NewExtrinsic(call)
+	// Get latest runtime version
+	rv, err := c.Conn.RPC.State.GetRuntimeVersionLatest()
+	if err != nil {
+		return nil, err
+	}
+
+	c.nonceLock.Lock()
+	defer c.nonceLock.Unlock()
+
+	nonce, err := c.nextNonce(&meta)
```

Comment:
On the EVM side we are storing nonce locally and acessing it with mutex to eliminate race conditions and node memorypool missmatch problems when fetching latest nonce. Should not we do the same here?

Reply:
It is kinda the same here except we first ask for the nonce in storage and then compare with latest account nonce from RPC call.
The nonce is still protected by nonce lock that way.